### PR TITLE
Support querying contract state for config in vanilla OCR2 and mercury

### PR DESCRIPTION
**Pull request by samsondav**

### CODE REVIEW

**Review by austinborn — COMMENTED** [review]
The `Verifier` contract looks great, although will this need to be added for any of the other contracts, like `OCR2Base.sol`?

File: go.mod [hunk 1]
```diff
@@ -381,4 +381,7 @@ replace (
 
 	// until merged upstream: https://github.com/hashicorp/go-plugin/pull/257
 	github.com/hashicorp/go-plugin => github.com/jmank88/go-plugin v0.0.0-20230604120638-7bb12ec27e75
+
+	github.com/smartcontractkit/libocr => /Users/sam/code/smartcontractkit/libocr
```
[the remark this answers is not in the document]
Comment (samsondav) [reply go.mod:385]: I need this PR to be merged before I can remove this https://github.com/smartcontractkit/offchain-reporting/pull/434

File: core/services/relay/evm/config_poller.go [hunk 2]
```diff
@@ -157,3 +212,62 @@ func (cp *configPoller) LatestBlockHeight(ctx context.Context) (blockHeight uint
...
+func (cp *configPoller) callReadConfigFromStore(ctx context.Context, expectChangedInBlock uint64) (cfg ocrtypes.ContractConfig, err error) {
+	changedInBlock, configDigest, err := cp.LatestConfigDetails(ctx)
+	if err != nil {
+		failedRPCContractCalls.WithLabelValues(cp.client.ConfiguredChainID().String(), cp.aggregatorContractAddr.Hex()).Inc()
+		return cfg, fmt.Errorf("failed to get latest config details: %w", err)
+	}
+	if configDigest == (ocrtypes.ConfigDigest{}) {
+		return cfg, nil
+	}
+	// FIXME: How to handle this case?
```
Comment (samsondav) [inline core/services/relay/evm/config_poller.go:241]: Don't merge til this is resolved

File: docs/CHANGELOG.md [hunk 3]
```diff
@@ -23,6 +23,15 @@ AllowSimplePasswords=true
...
 - To migrate on production builds, update the database password set in Database.URL to be 16 - 50 characters without leading or trailing whitespace. URI parsing rules apply to the chosen password - refer to [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986) for special character escape rules.
 
+### Added
+
+- OCR2 jobs now support querying the state contract for configurations if it has been deployed. 
+  This can help on chains such as BSC which "manage" state bloat by arbitrarily deleting logs 
+  older than a certain date. In this case, if logs are missing we will query the contract directly
+  and retrieve the latest config from chain state. Chainlink will perform no
+  extra RPC calls unless the contract has this feature explicitly enabled. On
+  chains that require this, nops may see an increase in RPC calls.
```
Comment (samsondav) [inline docs/CHANGELOG.md:33]: Changelog needs updating

File: core/services/relay/evm/functions/config_poller_test.go [hunk 4]
```diff
@@ -203,3 +204,27 @@ func setFunctionsConfig(t *testing.T, pluginConfig *functionsConfig.ReportingPlu
 		OffchainConfig:        offchainConfig,
 	}
 }
+
+func generateDefaultOCR2OnchainConfig(minValue *big.Int, maxValue *big.Int) []byte {
```
Comment (Fletch153) [inline core/services/relay/evm/functions/config_poller_test.go:208]: @samsondav any suggestions where we can put this utility function for building onchain configurations?

Comment (samsondav) [reply core/services/relay/evm/functions/config_poller_test.go:208]: If its used in tests it doesnt matter too much. You could put in `evmutils`

File: docs/CHANGELOG.md [hunk 5]
```diff
@@ -23,6 +23,17 @@ AllowSimplePasswords=true
...
+
+- OCR2 jobs now support querying the state contract for configurations if it has been deployed. 
+  This can help on chains such as BSC which "manage" state bloat by arbitrarily deleting logs 
+  older than a certain date. In this case, if logs are missing we will query the contract directly
+  and retrieve the latest config from chain state. Chainlink will perform no
+  extra RPC calls unless the job spec has this feature explicitly enabled. On
+  chains that require this, nops may see an increase in RPC calls. This can be
+  enabled for OCR2 jobs by specifying `configContractAddress` in the relay
+  config TOML.
+
```
Comment (jmank88) [inline docs/CHANGELOG.md:36]: Move up to `[dev]`?

File: core/internal/features/ocr2/features_ocr2_test.go [hunk 6]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
+	"github.com/smartcontractkit/libocr/bigbigendian"
```
Comment (jmank88) [inline core/internal/features/ocr2/features_ocr2_test.go:7]: nit/ unsorted imports

File: core/services/relay/evm/functions/config_poller_test.go [hunk 7]
```diff
@@ -205,3 +206,27 @@ func setFunctionsConfig(t *testing.T, pluginConfig *functionsConfig.ReportingPlu
...
```
Comment (jmank88) [inline core/services/relay/evm/functions/config_poller_test.go:210]: If you pass a `t`, then you can fail the test without disrupting the whole suite with `panic`.
```suggestion
func generateDefaultOCR2OnchainConfig(t *testing.T, minValue *big.Int, maxValue *big.Int) []byte {
```

File: core/internal/features/ocr2/features_ocr2_test.go [hunk 8]
```diff
@@ -722,4 +738,28 @@ juelsPerFeeCoinSource = """
 	assert.Equal(t, digestAndEpoch.Epoch, epoch)
 }
 
+func generateDefaultOCR2OnchainConfig(minValue *big.Int, maxValue *big.Int) []byte {
```
Comment (jmank88) [inline core/internal/features/ocr2/features_ocr2_test.go:741]: Could use a `T` here too?
```suggestion
func generateDefaultOCR2OnchainConfig(t *testing.T, minValue *big.Int, maxValue *big.Int) []byte {
```

File: core/services/relay/evm/functions/config_poller_test.go [hunk 9]
```diff
@@ -205,3 +207,27 @@ func setFunctionsConfig(t *testing.T, pluginConfig *functionsConfig.ReportingPlu
 		OffchainConfig:        offchainConfig,
 	}
 }
+
+func generateDefaultOCR2OnchainConfig(t *testing.T, minValue *big.Int, maxValue *big.Int) []byte {
+	serializedConfig := make([]byte, 0)
```
Comment (jmank88) [inline core/services/relay/evm/functions/config_poller_test.go:212]: just stylistic, but I'm surprised the linter doesn't complain about this one:
```suggestion
	var serializedConfig []byte
```

File: core/services/relay/evm/config_poller.go [hunk 10]
```diff
@@ -117,11 +162,14 @@ func (cp *configPoller) Replay(ctx context.Context, fromBlock int64) error {
...
+	latest, err := cp.destChainLogPoller.LatestLogByEventSigWithConfs(ConfigSet, cp.aggregatorContractAddr, 1, pg.WithParentCtx(ctx))
 	if err != nil {
-		// If contract is not configured, we will not have the log.
 		if errors.Is(err, sql.ErrNoRows) {
-			return 0, ocrtypes.ConfigDigest{}, nil
+			if cp.isConfigStoreAvailable() {
+				// Fallback to RPC call in case logs have been pruned and configStoreContract is available
+				return cp.callLatestConfigDetails(ctx)
+			}
+			err = nil // log not found means return zero config digest
```
Comment (jmank88) [inline core/services/relay/evm/config_poller.go:172]: Falling back in to the other error logic to achieve this make me nervous about future changes. Would it be more clear and less brittle to return directly from here:
```suggestion
			// log not found means return zero config digest
			return 0, ocrtypes.ConfigDigest{}, nil
```

**Review by jmank88 — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]

**Pull request merged**